iceoryx_posh/include/iceoryx_posh/popo/wait_set.hpp

```diff
@@ -73,71 +75,95 @@ class WaitSet
+    template <typename T,
+              typename EventType,
+              typename ContextDataType = internal::NoType_t,
+              typename = std::enable_if_t<std::is_enum<EventType>::value, void>>
+    cxx::expected<WaitSetError> attachEvent(T& eventOrigin,
+                                            const EventType eventType,
+                                            const EventCallback<T, ContextDataType>& eventCallback) noexcept;
 
     /// @brief attaches an event of a given class to the WaitSet.
     /// @param[in] eventOrigin the class from which the event originates.
     /// @param[in] eventId an arbitrary user defined id for the event
     /// @param[in] eventCallback a callback which should be assigned to the event
-    template <typename T>
-    cxx::expected<WaitSetError>
-    attachEvent(T& eventOrigin, const uint64_t eventId = 0U, const EventInfo::Callback<T>& eventCallback = {}) noexcept;
+    template <typename T, typename ContextDataType = internal::NoType_t>
+    cxx::expected<WaitSetError> attachEvent(T& eventOrigin,
+                                            const uint64_t eventId = 0U,
+                                            const EventCallback<T, ContextDataType>& eventCallback = {}) noexcept;
 
     /// @brief attaches an event of a given class to the WaitSet.
     /// @param[in] eventOrigin the class from which the event originates.
     /// @param[in] eventCallback a callback which should be assigned to the event
-    template <typename T>
-    cxx::expected<WaitSetError> attachEvent(T& eventOrigin, const EventInfo::Callback<T>& eventCallback) noexcept;
+    template <typename T, typename ContextDataType = internal::NoType_t>
+    cxx::expected<WaitSetError> attachEvent(T& eventOrigin,
+                                            const EventCallback<T, ContextDataType>& eventCallback) noexcept;
 
     /// @brief attaches a state of a given class to the WaitSet.
     /// @param[in] stateOrigin the class from which the state originates.
     /// @param[in] stateType the state specified by the class
     /// @param[in] id an arbitrary user defined id for the state
     /// @param[in] stateCallback a callback which should be assigned to the state
-    template <typename T, typename StateType, typename = std::enable_if_t<std::is_enum<StateType>::value>>
+    template <typename T,
+              typename StateType,
+              typename ContextDataType = internal::NoType_t,
+              typename = std::enable_if_t<std::is_enum<StateType>::value>>
     cxx::expected<WaitSetError> attachState(T& stateOrigin,
                                             const StateType stateType,
                                             const uint64_t id = 0U,
-                                            const EventInfo::Callback<T>& stateCallback = {}) noexcept;
+                                            const EventCallback<T, ContextDataType>& stateCallback = {}) noexcept;
```

Comment:
Maybe I overlooked it, but I couldn't find any test for `attachState` where you provide the `stateCallback`.